File: cli/src/agent_swarm/parsers.py
"""JSON output parsers for different CLI agents.

Each CLI has a different output format. This module normalizes them to a common schema.

Common event types (normalized):
- init: Agent started
- thinking: Agent reasoning (delta or complete)
- tool_use: Tool was called
- file_write: File was written/modified
- file_read: File was read
- bash: Bash command executed
- message: Agent message (delta or complete)
- result: Final result
- error: Error occurred
- raw: Unparsed output
"""

from datetime import datetime
from typing import Literal

AgentType = Literal["codex", "gemini", "cursor", "claude"]


def normalize_events(agent_type: AgentType, raw: dict) -> list[dict]:
    """Normalize an event from any agent to a list of common-schema events."""
    if agent_type == "codex":
        return _normalize_codex(raw)
    elif agent_type == "cursor":
        return _normalize_cursor(raw)
    elif agent_type == "gemini":
        return _normalize_gemini(raw)
    elif agent_type == "claude":
        return _normalize_claude(raw)

    timestamp = datetime.now().isoformat()
    return [{
        "type": raw.get("type", "unknown"),
        "agent": agent_type,
        "raw": raw,
        "timestamp": timestamp,
    }]


def normalize_event(agent_type: AgentType, raw: dict) -> dict:
    """Normalize an event and return the first normalized entry.

    Consumers that need multiple derived events (e.g., multiple tool_use blocks)
    should call `normalize_events` instead.
    """
    events = normalize_events(agent_type, raw)
    if events:
        return events[0]

    return {
        "type": raw.get("type", "unknown"),
        "agent": agent_type,
        "raw": raw,
        "timestamp": datetime.now().isoformat(),
    }


def _normalize_codex(raw: dict) -> list[dict]:
    """
    Normalize Codex CLI output.

    Codex events:
    - thread.started: {"type": "thread.started", "thread_id": "..."}
    - turn.started: {"type": "turn.started"}
    - item.completed: {"type": "item.completed", "item": {"type": "agent_message|tool_call", ...}}
    - turn.completed: {"type": "turn.completed", "usage": {...}}
    """
    event_type = raw.get("type", "unknown")
    timestamp = datetime.now().isoformat()

    if event_type == "thread.started":
        return [{
            "type": "init",
            "agent": "codex",
            "session_id": raw.get("thread_id"),
            "timestamp": timestamp,
        }]

    elif event_type == "turn.started":
        return [{
            "type": "turn_start",
            "agent": "codex",
            "timestamp": timestamp,
        }]

    elif event_type == "item.completed":
        item = raw.get("item", {})
        item_type = item.get("type")

        if item_type == "agent_message":
            return [{
                "type": "message",
                "agent": "codex",
                "content": item.get("text", ""),
                "complete": True,
                "timestamp": timestamp,
            }]

        elif item_type == "command_execution":
            # Codex runs shell commands via command_execution
            command = item.get("command", "")
            return [{
                "type": "bash",
                "agent": "codex",
                "tool": "command_execution",
                "command": command,
                "timestamp": timestamp,
            }]

        elif item_type == "tool_call":
            tool_name = item.get("name", "unknown")
            tool_args = item.get("arguments", {})

            # Detect file operations
            if tool_name in ("write_file", "create_file", "edit_file"):
                return [{
                    "type": "file_write",
                    "agent": "codex",
                    "tool": tool_name,
                    "path": tool_args.get("path", tool_args.get("file_path", "")),
                    "timestamp": timestamp,
                }]
            elif tool_name in ("read_file",):
                return [{
                    "type": "file_read",
                    "agent": "codex",
                    "tool": tool_name,
                    "path": tool_args.get("path", tool_args.get("file_path", "")),
                    "timestamp": timestamp,
                }]
            elif tool_name in ("shell", "bash", "execute"):
                return [{
                    "type": "bash",
                    "agent": "codex",
                    "tool": tool_name,
                    "command": tool_args.get("command", ""),
                    "timestamp": timestamp,
                }]
            else:
                return [{
                    "type": "tool_use",
                    "agent": "codex",
                    "tool": tool_name,
                    "args": tool_args,
                    "timestamp": timestamp,
                }]

    elif event_type == "turn.completed":
        usage = raw.get("usage", {})
        return [{
            "type": "result",
            "agent": "codex",
            "status": "success",
            "usage": {
                "input_tokens": usage.get("input_tokens", 0),
                "output_tokens": usage.get("output_tokens", 0),
            },
            "timestamp": timestamp,
        }]

    # Default passthrough
    return [{
        "type": event_type,
        "agent": "codex",
        "raw": raw,
        "timestamp": timestamp,
    }]


def _normalize_cursor(raw: dict) -> list[dict]:
    """
    Normalize Cursor Agent CLI output.

    Cursor events:
    - system/init: {"type": "system", "subtype": "init", "model": "...", "session_id": "..."}
    - user: {"type": "user", "message": {...}}
    - thinking: {"type": "thinking", "subtype": "delta|complete", "text": "..."}
    - assistant: {"type": "assistant", "message": {...}}
    - result: {"type": "result", "subtype": "success|error", "duration_ms": ...}
    """
    event_type = raw.get("type", "unknown")
    subtype = raw.get("subtype")
    timestamp = datetime.now().isoformat()

    if event_type == "system" and subtype == "init":
        return [{
            "type": "init",
            "agent": "cursor",
            "model": raw.get("model"),
            "session_id": raw.get("session_id"),
            "timestamp": timestamp,
        }]

    elif event_type == "thinking":
        return [{
            "type": "thinking",
            "agent": "cursor",
            "content": raw.get("text", ""),
            "complete": subtype == "complete",
            "timestamp": timestamp,
        }]

    elif event_type == "assistant":
        message = raw.get("message", {})
        content_blocks = message.get("content", [])
        events = []
        text_content = ""

        # Process all content blocks - create separate events for each tool_use
        for block in content_blocks:
            if block.get("type") == "text":
                text_content += block.get("text", "")
            elif block.get("type") == "tool_use":
                # Create a separate event for each tool_use block
                events.append({
                    "type": "tool_use",
                    "agent": "cursor",
                    "tool": block.get("name", "unknown"),
                    "args": block.get("input", {}),
                    "timestamp": timestamp,
                })

        # If there's text content and no tool_use blocks, or text comes after tool_use,
        # create a message event. If we already have tool_use events, we still create
        # a message event if there's text content.
        if text_content:
            events.append({
                "type": "message",
                "agent": "cursor",
                "content": text_content,
                "complete": True,
                "timestamp": timestamp,
            })

        # If no events were created (empty content blocks), return a message event anyway
        if not events:
            events.append({
                "type": "message",
                "agent": "cursor",
                "content": "",
                "complete": True,
                "timestamp": timestamp,
            })

        return events

    elif event_type == "result":
        return [{
            "type": "result",
            "agent": "cursor",
            "status": subtype or "success",
            "duration_ms": raw.get("duration_ms"),
            "timestamp": timestamp,
        }]

    elif event_type == "tool_result":
        return [{
            "type": "tool_result",
            "agent": "cursor",
            "tool": raw.get("tool_name", "unknown"),
            "success": raw.get("success", True),
            "timestamp": timestamp,
        }]

    # Default passthrough
    return [{
        "type": event_type,
        "agent": "cursor",
        "raw": raw,
        "timestamp": timestamp,
    }]


def _normalize_gemini(raw: dict) -> list[dict]:
    """
    Normalize Gemini CLI output.

    Gemini events:
    - init: {"type": "init", "timestamp": "...", "session_id": "...", "model": "..."}
    - message: {"type": "message", "role": "user|assistant", "content": "...", "delta": true|false}
    - result: {"type": "result", "status": "success|error", "stats": {...}}
    """
    event_type = raw.get("type", "unknown")
    timestamp = raw.get("timestamp", datetime.now().isoformat())

    if event_type == "init":
        return [{
            "type": "init",
            "agent": "gemini",
            "model": raw.get("model"),
            "session_id": raw.get("session_id"),
            "timestamp": timestamp,
        }]

    elif event_type == "message":
        role = raw.get("role", "assistant")
        if role == "assistant":
            return [{
                "type": "message",
                "agent": "gemini",
                "content": raw.get("content", ""),
                "complete": not raw.get("delta", False),
                "timestamp": timestamp,
            }]
        else:
            return [{
                "type": "user_message",
                "agent": "gemini",
                "content": raw.get("content", ""),
                "timestamp": timestamp,
            }]

    elif event_type in ("tool_call", "tool_use"):
        # Gemini uses tool_name/parameters, others use name/args
        tool_name_raw = raw.get("tool_name") or raw.get("name") or "unknown"
        tool_name = str(tool_name_raw)

        tool_args_raw = raw.get("parameters")
        if tool_args_raw is None:
            tool_args_raw = raw.get("args")
        tool_args = tool_args_raw if isinstance(tool_args_raw, dict) else {}
        tool_name_lower = tool_name.lower()

        # Detect file operations (handle both path and file_path keys)
        file_path = tool_args.get("file_path", tool_args.get("path", ""))

        if "write" in tool_name_lower and "file" in tool_name_lower:
            return [{
                "type": "file_write",
                "agent": "gemini",
                "tool": tool_name,
                "path": file_path,
                "timestamp": timestamp,
            }]
        elif "read" in tool_name_lower and "file" in tool_name_lower:
            return [{
                "type": "file_read",
                "agent": "gemini",
                "tool": tool_name,
                "path": file_path,
                "timestamp": timestamp,
            }]
        elif tool_name_lower in ("shell", "bash", "execute", "run_command"):
            return [{
                "type": "bash",
                "agent": "gemini",
                "tool": tool_name,
                "command": tool_args.get("command", ""),
                "timestamp": timestamp,
            }]

        return [{
            "type": "tool_use",
            "agent": "gemini",
            "tool": tool_name,
            "args": tool_args,
            "timestamp": timestamp,
        }]

    elif event_type == "result":
        stats = raw.get("stats", {})
        return [{
            "type": "result",
            "agent": "gemini",
            "status": raw.get("status", "success"),
            "duration_ms": stats.get("duration_ms"),
            "usage": {
                "total_tokens": stats.get("total_tokens", 0),
            },
            "timestamp": timestamp,
        }]

    # Default passthrough
    return [{
        "type": event_type,
        "agent": "gemini",
        "raw": raw,
        "timestamp": timestamp,
    }]


def _normalize_claude(raw: dict) -> list[dict]:
    """
    Normalize Claude Code CLI output.

    Similar to Cursor format (both use Anthropic's format).
    """
    # Claude Code uses similar format to Cursor
    events = _normalize_cursor(raw)

    # _normalize_cursor returns a list; mirror that but ensure agent is set correctly.
    for event in events:
        event["agent"] = "claude"
    return events


def parse_event(agent_type: AgentType, line: str) -> list[dict] | None:
    """Parse a single line of output into normalized events."""
    import json

    try:
        raw = json.loads(line)
        return normalize_events(agent_type, raw)
    except json.JSONDecodeError:
        return None

File: swarm/cli-ts/src/parsers.ts
export type AgentType = 'codex' | 'gemini' | 'cursor' | 'claude';

export function normalizeEvents(agentType: AgentType, raw: any): any[] {
  if (agentType === 'codex') {
    return normalizeCodex(raw);
  } else if (agentType === 'cursor') {
    return normalizeCursor(raw);
  } else if (agentType === 'gemini') {
    return normalizeGemini(raw);
  } else if (agentType === 'claude') {
    return normalizeClaude(raw);
  }

  const timestamp = new Date().toISOString();
  return [{
    type: raw.type || 'unknown',
    agent: agentType,
    raw: raw,
    timestamp: timestamp,
  }];
}

export function normalizeEvent(agentType: AgentType, raw: any): any {
  const events = normalizeEvents(agentType, raw);
  if (events.length > 0) {
    return events[0];
  }

  return {
    type: raw.type || 'unknown',
    agent: agentType,
    raw: raw,
    timestamp: new Date().toISOString(),
  };
}

function normalizeCodex(raw: any): any[] {
  const eventType = raw.type || 'unknown';
  const timestamp = new Date().toISOString();

  if (eventType === 'thread.started') {
    return [{
      type: 'init',
      agent: 'codex',
      session_id: raw.thread_id,
      timestamp: timestamp,
    }];
  } else if (eventType === 'turn.started') {
    return [{
      type: 'turn_start',
      agent: 'codex',
      timestamp: timestamp,
    }];
  } else if (eventType === 'item.completed') {
    const item = raw.item || {};
    const itemType = item.type;

    if (itemType === 'agent_message') {
      return [{
        type: 'message',
        agent: 'codex',
        content: item.text || '',
        complete: true,
        timestamp: timestamp,
      }];
    } else if (itemType === 'command_execution') {
      const command = item.command || '';
      return [{
        type: 'bash',
        agent: 'codex',
        tool: 'command_execution',
        command: command,
        timestamp: timestamp,
      }];
    } else if (itemType === 'tool_call') {
      const toolName = item.name || 'unknown';
      const toolArgs = item.arguments || {};

      if (toolName === 'write_file' || toolName === 'create_file' || toolName === 'edit_file') {
        return [{
          type: 'file_write',
          agent: 'codex',
          tool: toolName,
          path: toolArgs.path || toolArgs.file_path || '',
          timestamp: timestamp,
        }];
      } else if (toolName === 'read_file') {
        return [{
          type: 'file_read',
          agent: 'codex',
          tool: toolName,
          path: toolArgs.path || toolArgs.file_path || '',
          timestamp: timestamp,
        }];
      } else if (toolName === 'shell' || toolName === 'bash' || toolName === 'execute') {
        return [{
          type: 'bash',
          agent: 'codex',
          tool: toolName,
          command: toolArgs.command || '',
          timestamp: timestamp,
        }];
      } else {
        return [{
          type: 'tool_use',
          agent: 'codex',
          tool: toolName,
          args: toolArgs,
          timestamp: timestamp,
        }];
      }
    }
  } else if (eventType === 'turn.completed') {
    const usage = raw.usage || {};
    return [{
      type: 'result',
      agent: 'codex',
      status: 'success',
      usage: {
        input_tokens: usage.input_tokens || 0,
        output_tokens: usage.output_tokens || 0,
      },
      timestamp: timestamp,
    }];
  }

  return [{
    type: eventType,
    agent: 'codex',
    raw: raw,
    timestamp: timestamp,
  }];
}

function normalizeCursor(raw: any): any[] {
  const eventType = raw.type || 'unknown';
  const subtype = raw.subtype;
  const timestamp = new Date().toISOString();

  if (eventType === 'system' && subtype === 'init') {
    return [{
      type: 'init',
      agent: 'cursor',
      model: raw.model,
      session_id: raw.session_id,
      timestamp: timestamp,
    }];
  } else if (eventType === 'thinking') {
    return [{
      type: 'thinking',
      agent: 'cursor',
      content: raw.text || '',
      complete: subtype === 'completed',
      timestamp: timestamp,
    }];
  } else if (eventType === 'assistant') {
    const message = raw.message || {};
    const contentBlocks = message.content || [];
    const events: any[] = [];
    let textContent = '';

    for (const block of contentBlocks) {
      if (block.type === 'text') {
        textContent += block.text || '';
      } else if (block.type === 'tool_use') {
        events.push({
          type: 'tool_use',
          agent: 'cursor',
          tool: block.name || 'unknown',
          args: block.input || {},
          timestamp: timestamp,
        });
      }
    }

    if (textContent) {
      events.push({
        type: 'message',
        agent: 'cursor',
        content: textContent,
        complete: true,
        timestamp: timestamp,
      });
    }

    if (events.length === 0) {
      events.push({
        type: 'message',
        agent: 'cursor',
        content: '',
        complete: true,
        timestamp: timestamp,
      });
    }

    return events;
  } else if (eventType === 'result') {
    return [{
      type: 'result',
      agent: 'cursor',
      status: subtype || 'success',
      duration_ms: raw.duration_ms,
      timestamp: timestamp,
    }];
  } else if (eventType === 'tool_result') {
    return [{
      type: 'tool_result',
      agent: 'cursor',
      tool: raw.tool_name || 'unknown',
      success: raw.success !== false,
      timestamp: timestamp,
    }];
  } else if (eventType === 'tool_call' && subtype === 'completed') {
    const toolCall = raw.tool_call;

    if (toolCall?.shellToolCall) {
      const command = toolCall.shellToolCall.args?.command || '';
      return [{
        type: 'bash',
        agent: 'cursor',
        tool: 'shell',
        command: command,
        timestamp: timestamp,
      }];
    } else if (toolCall?.editToolCall) {
      const filePath = toolCall.editToolCall.args?.path || '';
      return [{
        type: 'file_write',
        agent: 'cursor',
        tool: 'edit',
        path: filePath,
        timestamp: timestamp,
      }];
    } else if (toolCall?.readToolCall) {
      const filePath = toolCall.readToolCall.args?.path || '';
      return [{
        type: 'file_read',
        agent: 'cursor',
        tool: 'read',
        path: filePath,
        timestamp: timestamp,
      }];
    }

    return [{
      type: 'tool_use',
      agent: 'cursor',
      tool: Object.keys(toolCall || {})[0] || 'unknown',
      timestamp: timestamp,
    }];
  }

  return [{
    type: eventType,
    agent: 'cursor',
    raw: raw,
    timestamp: timestamp,
  }];
}

function normalizeGemini(raw: any): any[] {
  const eventType = raw.type || 'unknown';
  const timestamp = raw.timestamp || new Date().toISOString();

  if (eventType === 'init') {
    return [{
      type: 'init',
      agent: 'gemini',
      model: raw.model,
      session_id: raw.session_id,
      timestamp: timestamp,
    }];
  } else if (eventType === 'message') {
    const role = raw.role || 'assistant';
    if (role === 'assistant') {
      return [{
        type: 'message',
        agent: 'gemini',
        content: raw.content || '',
        complete: !raw.delta,
        timestamp: timestamp,
      }];
    } else {
      return [{
        type: 'user_message',
        agent: 'gemini',
        content: raw.content || '',
        timestamp: timestamp,
      }];
    }
  } else if (eventType === 'tool_call' || eventType === 'tool_use') {
    const toolNameRaw = raw.tool_name || raw.name || 'unknown';
    const toolName = String(toolNameRaw);

    let toolArgsRaw = raw.parameters;
    if (toolArgsRaw === null || toolArgsRaw === undefined) {
      toolArgsRaw = raw.args;
    }
    const toolArgs = (typeof toolArgsRaw === 'object' && toolArgsRaw !== null) ? toolArgsRaw : {};
    const toolNameLower = toolName.toLowerCase();

    const filePath = toolArgs.file_path || toolArgs.path || '';

    if (toolNameLower.includes('write') && toolNameLower.includes('file')) {
      return [{
        type: 'file_write',
        agent: 'gemini',
        tool: toolName,
        path: filePath,
        timestamp: timestamp,
      }];
    } else if (toolNameLower.includes('read') && toolNameLower.includes('file')) {
      return [{
        type: 'file_read',
        agent: 'gemini',
        tool: toolName,
        path: filePath,
        timestamp: timestamp,
      }];
    } else if (['shell', 'bash', 'execute', 'run_command'].includes(toolNameLower)) {
      return [{
        type: 'bash',
        agent: 'gemini',
        tool: toolName,
        command: toolArgs.command || '',
        timestamp: timestamp,
      }];
    }

    return [{
      type: 'tool_use',
      agent: 'gemini',
      tool: toolName,
      args: toolArgs,
      timestamp: timestamp,
    }];
  } else if (eventType === 'result') {
    const stats = raw.stats || {};
    return [{
      type: 'result',
      agent: 'gemini',
      status: raw.status || 'success',
      duration_ms: stats.duration_ms,
      usage: {
        total_tokens: stats.total_tokens || 0,
      },
      timestamp: timestamp,
    }];
  }

  return [{
    type: eventType,
    agent: 'gemini',
    raw: raw,
    timestamp: timestamp,
  }];
}

function normalizeClaude(raw: any): any[] {
  const events = normalizeCursor(raw);
  for (const event of events) {
    event.agent = 'claude';
  }
  return events;
}

export function parseEvent(agentType: AgentType, line: string): any[] | null {
  try {
    const raw = JSON.parse(line);
    return normalizeEvents(agentType, raw);
  } catch {
    return null;
  }
}
